packages/@lwc/engine-core/src/framework/utils.ts

```diff
@@ -4,47 +4,21 @@
  * SPDX-License-Identifier: MIT
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
-import { ArrayPush, create, isArray, isFunction, isUndefined, keys, seal } from '@lwc/shared';
+import { create, isArray, isUndefined, keys, seal } from '@lwc/shared';
 import { StylesheetFactory, TemplateStylesheetFactories } from './stylesheet';
 import { RefVNodes, VM } from './vm';
 import { VBaseElement, VStatic } from './vnodes';
 
-type Callback = () => void;
-
-let nextTickCallbackQueue: Callback[] = [];
 export const SPACE_CHAR = 32;
 
 export const EmptyObject = seal(create(null));
 export const EmptyArray = seal([]);
 
-function flushCallbackQueue() {
-    if (process.env.NODE_ENV !== 'production') {
-        if (nextTickCallbackQueue.length === 0) {
-            throw new Error(
-                `Internal Error: If callbackQueue is scheduled, it is because there must be at least one callback on this pending queue.`
-            );
-        }
-    }
-    const callbacks: Callback[] = nextTickCallbackQueue;
-    nextTickCallbackQueue = []; // reset to a new queue
-    for (let i = 0, len = callbacks.length; i < len; i += 1) {
-        callbacks[i]();
-    }
-}
-
-export function addCallbackToNextTick(callback: Callback) {
-    if (process.env.NODE_ENV !== 'production') {
-        if (!isFunction(callback)) {
-            throw new Error(
-                `Internal Error: addCallbackToNextTick() can only accept a function callback`
-            );
-        }
-    }
-    if (nextTickCallbackQueue.length === 0) {
-        Promise.resolve().then(flushCallbackQueue);
-    }
-    ArrayPush.call(nextTickCallbackQueue, callback);
-}
+// For browser support, see: https://caniuse.com/mdn-api_queuemicrotask
+export const addCallbackToNextTick =
+    typeof queueMicrotask === 'function'
+        ? queueMicrotask
+        : (callback: VoidFunction) => Promise.resolve().then(callback);
```

Comment:
Technically this changes the implementation, but I don't believe we need all this code. `queueMicrotask` accomplishes the same thing.

Reply:
The previous implementation had a deterministic ordering since it used a queue array. That's not a problem with queueMicrotask, but the fallback behavior could be a problem if it sets up multiple independent Promise chains since that could be nondeterministic.

That said, the only place this is being used now is for the rehydration queue, and taking a look at that code I don't _think_ this change in behavior matters?

Reply:
`queueMicrotask` and `Promise.resolve().then()` should both also be deterministic. The callbacks should execute in order.

I guess the only thing that might be different is that someone _else_ might insert a microtask _between_ two callbacks, whereas in the old implementation they'd all be flushed together. I'm not sure it's very likely that this is observable.

Since it is observable, though, and since I'd like to get this in for 246, I'll back this change out and open another issue to track using `queueMicrotask`.

Reply:
https://github.com/salesforce/lwc/issues/3621